### PR TITLE
Bug 1716835: Hooking Image Registry internal service certificates

### PR DESCRIPTION
**Pull request by ricardomaraschini**
This patch makes the default image registry self signed certificates available to OpenShift API Server pods.

### CODE REVIEW

**Comment by ricardomaraschini** [comment]
@adambkaplan @dmage This is the easy fix, it should be enough to spark some discussions around the  subject. 

I don't understand why we do have to keep track of certificates generated by the ca operator, one would expect these to be automatically accepted by other pods willing to use any service within the cluster.

File: pkg/operator/operatorclient/interfaces.go [hunk 1]
```diff
@@ -5,4 +5,5 @@ const (
 	GlobalMachineSpecifiedConfigNamespace = "openshift-config-managed"
 	OperatorNamespace                     = "openshift-apiserver-operator"
 	TargetNamespace                       = "openshift-apiserver"
+	ImageRegistryNamespace                = "openshift-image-registry"
```
Comment (mfojtik) [inline pkg/operator/operatorclient/interfaces.go]: i hate these consts, they make grepping code nightmare, can we just inline them?

File: pkg/operator/workloadcontroller/workload_controller.go [hunk 2]
```diff
@@ -34,6 +34,7 @@ import (
 const (
 	workloadDegradedCondition = "WorkloadDegraded"
 	imageImportCAName         = "image-import-ca"
+	internalRegistryCAName    = "image-registry-certificates"
```
Comment (mfojtik) [inline pkg/operator/workloadcontroller/workload_controller.go]: same here

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go [hunk 3]
```diff
@@ -252,39 +252,112 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
+// OpenShift internal registry access, the latter is a custom config map that may
+// be configured by the user on image.config.openshift.io/cluster.
+func mergeImageRegistryCertificates(cfgCli openshiftconfigclientv1.ConfigV1Interface, cli coreclientv1.CoreV1Interface) (map[string]string, error) {
+	cas := make(map[string]string)
+
+	internalRegistryCAs, err := cli.ConfigMaps(operatorclient.ImageRegistryNamespace).Get(
+		internalRegistryCAName, metav1.GetOptions{},
+	)
+	if err != nil && !apierrors.IsNotFound(err) {
+		return nil, err
```
Comment (mfojtik) [inline pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: if somebody delete the config map, are we going to remove it from merged configmap or keep it?

Comment (mfojtik) [reply pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: if we don't know, it is probably event worthy

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go [hunk 4]
```diff
@@ -252,39 +252,112 @@ func syncOpenShiftAPIServer_v311_00_to_latest(c OpenShiftAPIServerOperator, orig
...
-	}
-	if len(imageConfig.Spec.AdditionalTrustedCA.Name) == 0 {
-		existingCM, err := client.ConfigMaps(operatorclient.TargetNamespace).Get(imageImportCAName, metav1.GetOptions{})
-		if apierrors.IsNotFound(err) {
+
+	casConfigMap, err := client.ConfigMaps(operatorclient.TargetNamespace).Get(
+		imageImportCAName, metav1.GetOptions{},
+	)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
```
Comment (mfojtik) [inline pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: nit: move this outside of this if

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go (hunk 3 above)
Comment (sttts) [reply pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: events here in reconciliation will tend to be repeatedly fired.

Comment (ricardomaraschini) [reply pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: @mfojtik we are merging two config maps, **1** and **2**, into another, **3**. 

1. If only **1** is deleted, **3** will have the content of **2**.
2. If only **2** is deleted, **3** will have the content of **1**.
3. If **1** and **2** exist, **3** will have the content of both merged(user provided ConfigMap has precedence).
4. If **1** and **2** don't exist, **3** ceases to exist.

File: pkg/operator/operatorclient/interfaces.go (hunk 1 above)
Comment (sttts) [reply pkg/operator/operatorclient/interfaces.go]: +1

Comment (ricardomaraschini) [reply pkg/operator/operatorclient/interfaces.go]: Removed.

File: pkg/operator/workloadcontroller/workload_controller.go (hunk 2 above)
Comment (ricardomaraschini) [reply pkg/operator/workloadcontroller/workload_controller.go]: Removed.

File: pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go (hunk 4 above)
Comment (adambkaplan) [reply pkg/operator/workloadcontroller/workload_controller_openshiftapiserver_v311_00.go]: @ricardomaraschini please address

**Comment by adambkaplan** [comment]
bump @sttts @mfojtik

**Pull request merged**